### PR TITLE
delete-invalid-drainer-config

### PR DESCRIPTION
**Pull request by calvix**
towards https://github.com/giantswarm/giantswarm/issues/11473

### CODE REVIEW

**Comment by calvix** [comment]
added  a code that waits untils drainer is deleted before recreating it

**Comment by xh3b4sd** [comment]
That practice is not good. We should initiate an action and cancel in case the result of the action takes too long for this reconciliation loop. The antipattern here is waiting and blocking. We should rather come back later and try again. Like a transactional state machine. You have several steps and try to reconcile them. Once you try and fail you give up and come back again in the next loop. You do not stick to your "failure" and block progress of other steps.

**Comment by calvix** [comment]
@xh3b4sd does this look better?

File: service/controller/resource/drainerinitializer/resource.go [hunk 1]
```diff
@@ -249,22 +249,32 @@ func (r *Resource) ensure(ctx context.Context, obj interface{}) error {
...
-				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found drainer config for ec2 instance %#q", *instance.InstanceId))
+				// if the cluster id or instance id does not match, delete the bad CR and recreate it
+				if dc.Labels[key.TagCluster] != key.ClusterID(&cl) || dc.Annotations[annotation.InstanceID] != *instance.InstanceId {
+					err = r.g8sClient.CoreV1alpha1().DrainerConfigs(cr.GetNamespace()).Delete(privateDNS, &metav1.DeleteOptions{})
+					if err != nil {
+						return microerror.Mask(err)
+					}
+					r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting drainer config with wrong cluster id or instance id for ec2 instance %#q", *instance.InstanceId))
+					// cancel resource to let deletion  happen
+					// the drainer will be recreated with proper details next loop
```
Comment (xh3b4sd) [inline service/controller/resource/drainerinitializer/resource.go:271]:
```suggestion
					// cancel resource to let deletion happen the drainer config will be
					// recreated with proper details next loop
```

File: service/controller/resource/drainerinitializer/resource.go [hunk 2]
```diff
@@ -249,22 +249,32 @@ func (r *Resource) ensure(ctx context.Context, obj interface{}) error {
...
 				if err != nil {
 					return microerror.Mask(err)
 				}
-
 			} else if err != nil {
 				return microerror.Mask(err)
 			} else {
-				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found drainer config for ec2 instance %#q", *instance.InstanceId))
+				// if the cluster id or instance id does not match, delete the bad CR and recreate it
+				if dc.Labels[key.TagCluster] != key.ClusterID(&cl) || dc.Annotations[annotation.InstanceID] != *instance.InstanceId {
```
Comment (xh3b4sd) [inline service/controller/resource/drainerinitializer/resource.go:264]: No show stopper, but I would like to put this logic in a separate function that has a descriptive name.

File: service/controller/resource/drainerinitializer/resource.go [hunk 3]
```diff
@@ -249,22 +249,32 @@ func (r *Resource) ensure(ctx context.Context, obj interface{}) error {
...
+					err = r.g8sClient.CoreV1alpha1().DrainerConfigs(cr.GetNamespace()).Delete(privateDNS, &metav1.DeleteOptions{})
+					if err != nil {
+						return microerror.Mask(err)
+					}
+					r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting drainer config with wrong cluster id or instance id for ec2 instance %#q", *instance.InstanceId))
```
Comment (xh3b4sd) [inline service/controller/resource/drainerinitializer/resource.go:269]:
```suggestion
					r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted leftover drainer config for ec2 instance %#q", *instance.InstanceId))
```

File: service/controller/resource/drainerinitializer/resource.go (hunk 1 above)
Comment (xh3b4sd) [reply service/controller/resource/drainerinitializer/resource.go:271]: Let's also please put our standard logging "canceling resource". We need this consistency for easy debugging.

File: service/controller/resource/drainerinitializer/resource.go (hunk 2 above)
Comment (calvix) [reply service/controller/resource/drainerinitializer/resource.go]: done

File: service/controller/resource/drainerinitializer/resource.go [hunk 4]
```diff
@@ -249,22 +249,34 @@ func (r *Resource) ensure(ctx context.Context, obj interface{}) error {
...
 				if err != nil {
 					return microerror.Mask(err)
 				}
-
 			} else if err != nil {
 				return microerror.Mask(err)
 			} else {
-				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found drainer config for ec2 instance %#q", *instance.InstanceId))
+				// if the cluster id or instance id does not match, delete the bad CR and recreate it
+				if key.IsWrongDrainerConfig(dc, key.ClusterID(&cl), *instance.InstanceId) {
```
Comment (xh3b4sd) [inline service/controller/resource/drainerinitializer/resource.go:264]: A private function in here would have worked as well.

**Pull request merged**